tests/e2e-pw/specs/merchant/merchant-subscriptions-settings.spec.ts

```diff
@@ -0,0 +1,24 @@
+/**
+ * External dependencies
+ */
+import test, { expect } from 'playwright/test';
+import { describeif, useMerchant } from '../../utils/helpers';
+import { shouldRunSubscriptionsTests } from '../../utils/constants';
+import { goToWooCommerceSettings } from '../../utils/merchant-navigation';
+
+describeif( shouldRunSubscriptionsTests )(
+	'WooCommerce > Settings > Subscriptions',
+	() => {
+		useMerchant();
+		test( 'Merchant should be able to load WooCommerce Subscriptions settings tab', async ( {
+			page,
+		} ) => {
+			await goToWooCommerceSettings( page, 'subscriptions' );
+			expect(
+				page.locator( 'a.nav-tab-active', {
+					hasText: 'Subscriptions',
+				} )
+			).toBeVisible();
```

Comment:
Although it works just fine, I think we should try to use the `getByRole` and `getByText` functions, which matches better the [preferred way](https://github.com/Automattic/woocommerce-payments/blob/e74afe26af6069a940499027e84108bfb1dc4d39/tests/e2e-pw/README.md?plain=1#L36) of searching for elements.

This is an alternative for checking if it's visible and if it contains the `.nav-tab-active` class, so we ensure it's marked as selected.

```suggestion
			const menuItem = page.getByRole( 'link', {
				name: 'Subscriptions',
				exact: true,
			} );

			await expect( menuItem ).toBeVisible();
			await expect( menuItem ).toHaveClass( /nav-tab-active/ );
```

I like to use the "Pick locator" tool from the UI mode (on the left of the Locator tab) to generate a good locator.

<img width="299" alt="image" src="https://github.com/user-attachments/assets/c4964031-26f8-4577-8e3a-c217cd4fa117" />

Reply:
💯 using user-based locators rather than implementation-based locators makes our tests more resilient to changes (e.g. if the nav element classname changes in WC core).

In this case, there are 2 `link` elements with the text `Subscriptions` (there is another in the admin sidebar menu), but we can scope to the `main` section of the page.

```suggestion
			const menuItem = page.getByRole( 'main' ).getByRole( 'link', {
				name: 'Subscriptions',
				exact: true,
			} );
			await expect( menuItem ).toBeVisible();

			// An alternative way to verify the subscriptions menu page is active, avoiding the active tab classname.
			const heading = await page.getByRole( 'heading', {
				name: 'Subscriptions',
			} );
			await expect( heading ).toBeVisible();
```

Reply:
Text can change too :) Changing H2 to DIV can also break things. If you can't adapt to the change, nothing can prevent tests to break down. There are many write-ups out there that say:

- Don't use texts
- Don't use CSS selectors
- Don't use ID's 

etc, the only thing I couldn't find is "Don't use test ID's" :D But I'll appreciate if you can share some publications that explain it in-depth. 

Thanks for the suggestions!

Reply:
Okay, I think we suggest using strings because the selectors are not all in our control, WC Core can change selectors which can break our tests, but we control the strings we produce. Gotcha. If we use the selectors we own, that would pose less danger to our tests though, but strings are the most reliable pointers.

Reply:
> strings are the most reliable pointers.

To explain my thoughts on this further: locating elements by role, text or label is more reliable because these more closely represent how users and assistive technology perceive the page, which is what we are testing in e2e tests (user flows).

I find it helpful to consider the mental model presented in this "Golden Rule of Assertions" when writing e2e tests (also discussed in this P2 post pcShBQ-1pz-p2):

> *A test must fail if, and only if, the intention behind the system is not met.*
> – https://www.epicweb.dev/the-golden-rule-of-assertions

In this case, we're testing the flow "Merchant should be able to load WooCommerce Subscriptions settings tab". So what we're testing is the presence of the settings tab for Subscriptions, and the presence of the tab content. The check for a "Subscriptions" `link` role should only break if the user flow changes (e.g. the tab label changes, or is missing) and not if WC change classnames, as you mention.

I've also suggested checking for the presence of the "Subscriptions" [heading role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/heading_role) (can be H1-H6) rather than an active tab classname (`.nav-tab-active`) for a similar reason – how can we test the intention of the test and _avoid_ testing implementation details not required for the user flow.

Apologies for the noisy drive-by review 😅 but I am hoping to encourage us to try to make our e2e tests more resilient since they are notoriously flakey.

Note that in Puppeteer, testing in this way was not as easy to achieve. Playwright, however, actively encourages and promotes developers to locate elements in this more resilient way:

> Playwright comes with multiple built-in locators. To make tests resilient, we recommend prioritizing user-facing attributes and explicit contracts such as [page.getByRole()](https://playwright.dev/docs/api/class-page#page-get-by-role).
> – https://playwright.dev/docs/locators#locating-elements

Reply:
Okay, thanks both, that was very informative. I tried to convert what was written in the Puppeteer test directly, I'll take that into account for further ports.